File: Java/src/Maximum Gap.java
public class Solution {
public int maximumGap(int[] num) {
    if (num == null || num.length < 2) return 0;

    int min = num[0];
    int max = num[0];
    for (int i:num) {
        if (max < i){
            max = i;
        }
        else if (min > i){
            min = i;
        }
    }
    
    int average = (int)Math.ceil((double)(max - min)/(num.length - 1));
    int[] mins = new int[num.length - 1]; 
    int[] maxs = new int[num.length - 1]; 
    Arrays.fill(mins, Integer.MAX_VALUE);
    Arrays.fill(maxs, Integer.MIN_VALUE);

    for (int i:num) {
        if (i == min || i == max)
            continue;
        int index = (i - min) / average; 
        mins[index] = Math.min(i, mins[index]);
        maxs[index] = Math.max(i, maxs[index]);
    }

    int ret = Integer.MIN_VALUE;
    int previous = min;
    for (int i = 0; i < num.length - 1; i++) {
        if (mins[i] == Integer.MAX_VALUE && maxs[i] == Integer.MIN_VALUE)
            continue;
        ret = Math.max(ret, mins[i] - previous);
        previous = maxs[i];
    }
    ret = Math.max(ret, max - previous);
    return ret;
}
}

File: Python/Maximum Gap.py
class Solution:
    # @param num, a list of integer
    # @return an integer
    def maximumGap(self, num):
        if len(num) < 2: return 0
        minNum = min(num)
        maxNum = max(num)
        mins = [float('inf')] * (len(num)-1)
        maxs = [-1] * (len(num) - 1)
        
        average = (maxNum - minNum-1)//(len(num) - 1) + 1
        
        for i in num:
            if i == minNum or i == maxNum:
                continue
            index = (i-minNum) // average
            mins[index] = min(mins[index], i)
            maxs[index] = max(maxs[index], i)
        
        pre = minNum
        ret = -1
        for i in range(len(num) - 1):
            if mins[i] == float('inf') and maxs[i] == -1:
                continue
            ret = max(ret, mins[i] - pre)
            pre = maxs[i]
        ret = max(ret, maxNum - pre)
        return ret
            

        
        